### ISSUE TITLE
improve search speed

### ISSUE THREAD

**Issue by jaap-karssenberg** [jaap-karssenberg/zim-desktop-wiki#64]
Page search is very slow compared to e.g. unix grep. 

Improve this by actually using grep as subprocess, or similar low-level read function to find "candidates". Only evaluate full query on these "candidates".

**Comment by Osndok** [comment]
I didn't really have time to explore this in depth (as I could have just been connecting it wrong), but I was surprised find that the search callback (that presents the items found so far and allows the search to be cancelled) might actually be slowing down the search considerably. Again, **NOT** 100% sure...

**Comment by quesada** [comment]
Ripgrep seems to be the one to beat. Maybe a plugin that does a system call to rg (instead of grep; or as an option, the same way we have 'import' and 'scrot' for screenshots) and displays the information as is now would work great and be a relatively low hanging fruit.

**Comment by Self-Perfection** [comment]
Prefiltering with tools like grep o ripgrep overcomes CPU bottleneck, but there still might be IO bottleneck on slow storage, like rotary HDDs. Therefore I find myself usually searching in zim notebook with `git grep --cached` on my latptop. It does not need to read couple of thousand txt files from HDD and performs in less than a second. And for comparison grep runs ~20s if notebook is not cached in RAM.

mercurial also has grep tool, though I'm not sure if it can give similar speedup.

**Comment by introt** [comment]
I've [ported](https://github.com/introt/zim-desktop-wiki/commit/28ed20e65c8157baea607587d1f027d23dc5d924) Osndok's patch to the latest version of Zim. It should work on Windows, as long as you've got `grep` in your path; let me know if it doesn't.

It should be noted that this version **replaces** the "native" implementation instead of complementing `_process_content` by prefiltering. This means your search string is used as grep patterns*; queries aren't processed. As an avid grep enjoyer, I'd like to present that as an option in the search dialog.

As for other options, making the grep base command (and thus executable) user editable in addition to plugin-style dependency checking should accommodate both `ripgrep` and `git grep --cached` users.

Not sure what's the next course of action towards a true implementation from here - @jaap-karssenberg should "grep search" be implemented in `zim.search` or as a plugin? Should "regular search" be made extendable, or replaceable by moving it into a plugin?

*multiple patterns are supported as-is - just copy and paste a newline ![image](https://user-images.githubusercontent.com/26069787/147549453-970cbfd9-163c-492f-a667-06471c2fe326.png)

**Comment by jaap-karssenberg** [comment]
I would prefer this to be a plugin, mostly because I don't want extensive
configuration options for this in the main preferences, and it could
potentially be multiple plugins to choose from.

Implementation wise I think the logic in "SearchSelection" need a big
overhaul. Would like to split a "Searcher" and "Selection" object (or
similar names) and have the searcher be build up from "search providers"
tied to keywords. That way the searcher can be extendable and plugins can
add keywords as well as overload default keywords.

However I would not want to block this improvement on such a full overhaul
of the module. Maybe an in between step can be made where this moves in a
plugin but with a "dirty" API hook.

Regards,

Jaap

**Comment by introt** [comment]
Agreed; `grep` and other third-party-tool calling search engines don't belong in the main app.

On the topic of improving search speed "natively", the Instant Search plugin seems to have this in the bag. I'm currently discussing licensing with the dev(s) over at e3rd/zim-plugin-instantsearch#48.

Not only is their implementation faster, it also solves #1138 (not by parsing the files, but by including possible formatting in the search regex - brilliant!).

If the search module is going to get overhaul'd, it'd make sense to incorporate (parts of) the plugin and provide API for eg. the Jump To (#189) and Link (#487) dialogs, whose auto-completion is currently implemented via a single SQL SELECT.

**Comment by AnonJohn** [comment]
Raising this thread to offer two reports: 1. 0.75 search seems considerably slower to me (windows, linux). 2. The instant search plugin seems to be extremely fast. Interface is not as good as native, but speed is amazing. If it could in fact be brought in that would seem a nice improvement. From what I can tell, the remaining obstacle is getting [@jtherr](https://github.com/jtherrmann) to sign off on the license change to GPL-2 for his contributions?

Since the plugin author seemed reluctant to make the request, would a direct ask potentially be the easy solution? (@introt)

**Comment by introt** [comment]
@AnonJohn haven't heard from @jtherr but main blocker is person-hours - I'm currently unavailable to undertake a task of this scale

**Comment by introt** [comment]
I reached out via email, the plugin's code is now fully GPLv2: https://github.com/jtherrmann/zim-plugin-instantsearch/pull/1

**Comment by maclomhair** [comment]
I'd like to propose a different approach to improving search speed: using sqlite's [full-text search functionality](https://www.sqlite.org/fts5.html), one could use the already existing index each zim wiki has and potentially outperform any grep-like variant. Do you think this is a viable and/or reasonable option?

**Comment by introt** [comment]
We don't store full text inside the index[^1], so unless we start doing that, it's not a viable option. And if we did, it'd mean even slower indexing each time someone starts a wiki with an empty cache, as some people, distro maintainers included, think that the cache should be deleted weekly. Could be somewhat faster[^2] on spinning rust after indexing though.

[^1]: unlike eg. Jopplin: https://joplinapp.org/spec/architecture/
[^2]: the Instant Search plugin is quite deserving of the name with ~100k lines in 10k pages on a SSD

**Comment by maclomhair** [comment]
From what I understand, full-text doesn't need to be stored in the index table to make use of the full-text search features of sqlite (the feature seems to be called "contentless tables" in their [docs](https://www.sqlite.org/fts5.html#external_content_and_contentless_tables)) - but building the cache would become much slower due to the creation of a tokenized reverse index, that's for sure.

That'd mean such a feature - if ever implemented - should be off by default and configurable in the settings, I guess, and improving speed of the "traditional" grep-like full-text search would remain of interest?

**Comment by introt** [comment]
The mention of Joplin was a bit misleading - what I meant was that the existing index is quite bare bones[^1], and in its current form can mainly assist in full-text search by providing the paths to wiki files.

However, given that its creation necessitates fully reading each file, building the search cache simultaneously might be just as quick when the I/O speed is the main bottleneck. In addition to people with local hard drives, a local cache could benefit those who store their wikis on their NAS or in the cloud.

If someone's willing to work on such feature, I'm happy to help get it merged[^3]. Improving Zim's indexing speed on the side would be very welcome.

[^1]: and needs optimization as-is. IIRC it fully parses each page "just" to collect links and tags, taking around 100 to 1000 times longer than GNU find + grep[^2]. It's a sane approach, but doesn't scale well.

[^2]: the GNU find + grep combo doesn't parse floating links which require knowing which .txt files are valid wiki pages, but that shouldn't take that much longer - Zim could eg. first read just the n first bytes of each file to check for the wiki header, write those into the database, and then do a second round on the wiki pages to look for links, tags etc., skipping the headers.

[^3]: Personally, if I find the time, I'm going to work on integrating the grep-like (regex-based) Instant Search plugin, because most of the work is already done - the code exists and has been tested, is fast enough (near instant at my scale), handles markup inside words etc. - and the approach might be easier to maintain.

**Comment by maclomhair** [comment]
Thank you for your supportive words! I have taken some time and attempted such a feature (pull request #2489 ). I was able to introduce most functionality as a plugin, but couldn't figure out how to get around modifying zim's core code in some locations. I'd gladly accept your help in getting these patches into a shape fit for merging!

**Comment by TwiggyWan** [comment]
Hey @maclomhair thanks for #2489 

I try to test it (`./zim.py` from latest develop commit) but get `sqlite version 3.43.0 or higher - Failed`.
I run zim.py from a virtualenv in which i have installed sqlite 3.46.0 and confirmed it was installed ok with  `python3 -c 'import sqlite3; print(sqlite3.sqlite_version_info)' `

Should I install sqlite3 differently?

I went through PLUGIN_WRITING.md already but did not find an answer.

**Comment by maclomhair** [comment]
Now that the indexed search is merged and released, I was wondering wether it's time to mark this issue as resolved? I've now used it more and more regularly even on pretty big wikis and indeed searching the entire wiki is now almost instantaneous, except for the first query after opening the notebook.

**Comment by AnonJohn** [comment]
Agree that this may be done. Search is absolutely brilliant (on linux, sorting dependencies on Windows). This plugin is fantastic and makes zim so much more usable for me (with my 10+ years of notes). TY so much for those who did the work.

**Comment by jaap-karssenberg** [comment]
Yes, the problem this issue describes is solved by the plugin. However, I want to keep the ticket open a bit longer to also address search speed without the plugin. Doing some refactoring in the search code to make extensions like this more flexible and make the basics better.
